soc/usbexecd.c

```diff
@@ -0,0 +1,884 @@
+
+    CMD_ACK_CONTROL = 0x20,
+    CMD_ACK_STDIN = 0x21,
+    CMD_ACK_STDOUT = 0x22,
+    CMD_ACK_STDERR = 0x23,
+
+    CMD_CLOSE_CONTROL = 0x30,
+    CMD_CLOSE_STDIN = 0x31,
+    CMD_CLOSE_STDOUT = 0x32,
+    CMD_CLOSE_STDERR = 0x33,
+};
+
+#define debug(args...)  syslog(LOG_DEBUG, args)
+#define info(args...)   syslog(LOG_INFO, args)
+#define error(args...)  syslog(LOG_ERR, args)
+#define fatal(args...) ({ \
+    syslog (LOG_CRIT, args); \
+    exit(1); \
+})
+
+int sock = 0;
+int epfd   = 0;
+int sigfd  = 0;
+
+#define PIPE_BUF 4096
+#define MAX_CTRL_ARGS 255
+#define MAX_WRITE_LEN 255
+
+enum Roles {
+    ROLE_CTRL = 0,
+    ROLE_STDIN = 1,
+    ROLE_STDOUT = 2,
+    ROLE_STDERR = 3,
+};
+
+typedef struct {
+    uint8_t id;
+    uint8_t role;
+    int events;
+    bool eof;
+    int credit;
+    int fd;
+    int startpos;
+    int endpos;
+    int bufcount;
+    char buffer[PIPE_BUF];
+} pipebuf_t;
+
+typedef struct {
+    int pid;
+    pipebuf_t ctrl;
+    pipebuf_t stdin;
+    pipebuf_t stdout;
+    pipebuf_t stderr;
+} procinfo_t;
+
+#define N_PROC 256
+procinfo_t* processes[N_PROC];
+
+void child(int ctrl, int stdin, int stdout, int stderr);
+void pipebuf_out_ack(pipebuf_t* pb, size_t acksize);
+void modify_pipebuf_epoll(pipebuf_t* pb, int operation);
+void add_pipebuf_epoll(pipebuf_t* pb);
+void delete_pipebuf_epoll(pipebuf_t* pb);
+void pipebuf_interal_write(pipebuf_t* pb, uint8_t *data);
+int write_from_pipebuf(pipebuf_t *pb, int fd, int len);
+void pipebuf_out_to_internal_buffer(pipebuf_t* pb, int len);
+void pipebuf_out_is_writable(pipebuf_t* pb);
+
+// Helper function to write a packet header to the domain socket
+// Param: cmd - which command in the Commands enum to send
+// Param: id - which process is sending this
+// Param: arg - an argument for the command (like ACK)
+// Param len - the length of the data being sent following this packet
+void send_header(uint8_t cmd, uint8_t id, uint8_t arg, uint8_t len) {
+    // Create the packet
+    uint8_t buf[4] = {cmd, id, arg, len};
+    // Write the packet to the socket
+    int r = write(sock, buf, sizeof(buf));
+    // Ensure it wrote all of the bytes
+    if (r != sizeof(buf)) {
+        fatal("send_header failed: %s", strerror(errno));
+    }
+}
+
+void read_until(int fd, void *buf, int len) {
+
+    int total_read = 0, single_read = 0;
+
+    while (total_read < len) {
+
+        single_read = read(fd, &(((uint8_t *)buf)[total_read]), len - total_read);
+
+        // If we read nothing and the socket would have blocked
+        if (single_read < 0 && (errno == EAGAIN || errno == EWOULDBLOCK)) {
+            // There is nothing to be read right now, try again
+            continue;
+        }
+        // We have read more than zero bytes sucessfully
+        else if (single_read > 0) {
+
+            // Increment the amount we have read
+            total_read += single_read;
+        }
+        else {
+            fatal("An error occured while reading child into pipebuf.");
```

Comment:
Error message doesn't reflect how this function is used; should also print `strerror(errno)` for debugging.